pkg/commands/compute/build.go

```diff
@@ -168,13 +195,98 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
 		})
 		return err
 	}
 
+	// FIXME: When we remove feature flag, put in ability to disable metadata.
+	if c.enableMetadata /*|| !disableWasmMetadata*/ {
+		endTime := time.Now()
+		runtime.ReadMemStats(&memAfter)
+
+		// FIXME: Once #1013 and #1016 merged, integrate granular disabling.
+		args := []string{
+			"metadata", "add", "bin/main.wasm",
+			fmt.Sprintf("--language=CLI_%s: %s", revision.AppVersion, strings.ToUpper(language.Name)),
+			fmt.Sprintf("--processed-by=CLI_BuildInfoMemoryHeapAlloc=%d", memAfter.HeapAlloc-memBefore.HeapAlloc),
+			fmt.Sprintf("--processed-by=CLI_BuildInfoTime=%s", endTime.Sub(startTime)),
+			fmt.Sprintf("--processed-by=CLI_MachineInfoOS=%s", runtime.GOOS),
+			fmt.Sprintf("--processed-by=CLI_MachineInfoArch=%s", runtime.GOARCH),
+			fmt.Sprintf("--processed-by=CLI_MachineInfoCompiler=%s", runtime.Compiler),
+			fmt.Sprintf("--processed-by=CLI_MachineInfoGoVersion=%s", runtime.Version()),
+			fmt.Sprintf("--processed-by=CLI_MachineInfoCPUs=%d", runtime.NumCPU()),
+		}
+
+		if c.Manifest.File.ClonedFrom != "" {
+			args = append(args, fmt.Sprintf("--processed-by=CLI_PackageInfoClonedFrom=%s", c.Manifest.File.ClonedFrom))
+		}
+
+		args = append(args, fmt.Sprintf("--processed-by=CLI_ScriptsDefaultBuildUsed=%t", language.DefaultBuildScript()))
+
+		if c.Manifest.File.Scripts.Build != "" {
+			args = append(args, fmt.Sprintf("--processed-by=CLI_ScriptsBuild=%s", c.Manifest.File.Scripts.Build))
+		}
+		if len(c.Manifest.File.Scripts.EnvVars) > 0 {
+			args = append(args, fmt.Sprintf("--processed-by=CLI_ScriptsEnvVars=%s", c.Manifest.File.Scripts.EnvVars))
+		}
+		if c.Manifest.File.Scripts.PostInit != "" {
+			args = append(args, fmt.Sprintf("--processed-by=CLI_ScriptsPostInit=%s", c.Manifest.File.Scripts.PostInit))
+		}
+		if c.Manifest.File.Scripts.PostBuild != "" {
+			args = append(args, fmt.Sprintf("--processed-by=CLI_ScriptsPostBuild=%s", c.Manifest.File.Scripts.PostBuild))
+		}
+
+		for k, v := range language.Dependencies() {
+			args = append(args, fmt.Sprintf("--sdk=%s=%s", k, v))
+		}
+
+		err = c.Globals.ExecuteWasmTools(wasmtools, args)
+		if err != nil {
+			return err
+		}
+
+		if c.showMetadata {
+			// gosec flagged this:
+			// G204 (CWE-78): Subprocess launched with variable
+			// Disabling as the variables come from trusted sources.
+			// #nosec
+			// nosemgrep
+			command := exec.Command(wasmtools, "metadata", "show", "bin/main.wasm")
+			wasmtoolsOutput, err := command.Output()
+			if err != nil {
+				return fmt.Errorf("failed to execute wasm-tools metadata command: %w", err)
+			}
+			text.Info(out, "Below is the metadata attached to the Wasm binary")
+			text.Break(out)
+
+			fmt.Fprintln(out, string(wasmtoolsOutput))
+			text.Break(out)
+		}
+	} else {
+		// NOTE: If not collecting metadata, just record the CLI version.
+		args := []string{
+			"metadata", "add", "bin/main.wasm",
+			fmt.Sprintf("--language=CLI_%s: %s", revision.AppVersion, strings.ToUpper(language.Name)),
```

Comment:
This feels like an odd way to encode the CLI version.  When I run `wasm-tools metadata show main.wasm` for a Rust program I get:

```
module:
    language:
        Rust
    processed-by:
        clang: 15.0.6
        rustc: 1.71.0 (8ede3aae2 2023-07-12)
```

with the code here I think that would appear as:

```
module:
    language:
        Rust
        CLI_10.4.0: Rust
    processed-by:
        clang: 15.0.6
        rustc: 1.71.0 (8ede3aae2 2023-07-12)
```

whereas something like this makes more sense IMO:

```
module:
    language:
        Rust
    processed-by:
        clang: 15.0.6
        rustc: 1.71.0 (8ede3aae2 2023-07-12)
        fastly: 10.4.0 (Rust)
```

Looking at `wasm-tools metadata add --help`:

```
      --language <NAME>
          Add a programming language to the producers section

      --processed-by <NAME=VERSION>
          Add a tool and its version to the producers section

      --sdk <NAME=VERSION>
          Add an SDK and its version to the producers section
```

A few things stand out to me.  `language` appears to just freeform text and not key-value like `processed-by` and `sdk`.

There doesn't appear to be a generic metadata section, so stuffing telemetry into `processed-by` feels a little weird to me, but without any better option it seems ok.  I'd consider writing it as JSON though, so that you're only using one key to do it and it's easily removed.

Reply:
Done.